### PR TITLE
feat(schema): provide type-level documentation snippets

### PR DESCRIPTION
**Pull request by keynslug**
For stuff like `duration()`, `bytesize()` and `secret()`, etc.

The latter will look like this in schema dumps:
```
  ...
  {
    "typedoc" : "A string holding some sensitive information, such as a password. When secret starts with <code>file://</code>, the rest of the string is interpreted as a path to a file containing the secret itself: whole content of the file except any trailing whitespace characters is considered a secret value.",
    "type" : "Secret",
    "text" : "password",
    "hash" : "V-authorization-S-authorization-sources-S-ldap-password",
    "doc" : "The password associated with the bridge, used for authentication with the external database."
  },
  ...
```

Fixes [EMQX-11579](https://emqx.atlassian.net/browse/EMQX-11579).

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [x] Schema changes are backward compatible

### CODE REVIEW

File: scripts/schema-dump-reformat.escript [hunk 1]
```diff
@@ -18,7 +18,8 @@ main(_) ->
     halt(1).
 
 reformat(Json) ->
-    emqx_conf:reformat_schema_dump(fix(Json)).
```
Comment (zmstone) [inline scripts/schema-dump-reformat.escript:21]: this script can be deleted.

File: apps/emqx_conf/src/emqx_conf_schema_types.erl [hunk 2]
```diff
@@ -165,37 +167,37 @@ readable("duration()") ->
...
         dashboard => #{type => duration},
-        docgen => #{type => "String", example => <<"12m">>}
+        docgen => #{type => "Duration", example => <<"12m">>, desc => ?DESC(duration)}
     };
 readable("timeout_duration_s()") ->
     #{
         swagger => #{type => string, example => <<"1h">>},
         dashboard => #{type => duration},
-        docgen => #{type => "String", example => <<"1h">>}
+        docgen => #{type => "Duration(s)", example => <<"1h">>, desc => ?DESC(duration)}
```
Comment (zmstone) [inline apps/emqx_conf/src/emqx_conf_schema_types.erl:194]: the `_s` or `_ms` types are only different internally.
Should be the same user-facing type. i.e. users should not need to know if this is converted to seconds or to milliseconds internally.

File: rel/i18n/emqx_conf_schema_types.hocon [hunk 3]
```diff
@@ -0,0 +1,12 @@
+emqx_conf_schema_types {
+
+    duration.desc:
+    """A string that represents a time duration, for example: <code>10s</code>, <code>2.5m</code>, <code>1h30m</code>, <code>1W2D</code>, or <code>2345ms</code>, which is the smallest unit. When precision is specified, finer portions of the duration may be ignored: writing <code>1200ms</code> for <code>Duration(s)</code> is equivalent to writing <code>1s</code>. It doesn't matter if units are in upper or lower case."""
```
Comment (zmstone) [inline rel/i18n/emqx_conf_schema_types.hocon:4]:
```suggestion
    """A string that represents a time duration, for example: <code>10s</code>, <code>2.5m</code>, <code>1h30m</code>, <code>1W2D</code>, or <code>2345ms</code>, which is the smallest unit. When precision is specified, finer portions of the duration may be ignored: writing <code>1200ms</code> for <code>Duration(s)</code> is equivalent to writing <code>1s</code>. The unit part is case-insensitive."""
```

File: rel/i18n/emqx_conf_schema_types.hocon [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+    bytesize.desc:
+    """A string that represents a number of bytes, for example: <code>10B</code>, <code>640kb</code>, <code>4MB</code>, <code>1GB</code>. Units are interpreted as powers of 1024, and the unit part is case-insensitive."""
+
+    secret.desc:
+    """A string holding some sensitive information, such as a password. When secret starts with <code>file://</code>, the rest of the string is interpreted as a path to a file containing the secret itself: whole content of the file except any trailing whitespace characters is considered a secret value."""
```
Comment (zmstone) [inline rel/i18n/emqx_conf_schema_types.hocon:10]:
```suggestion
    """A string holding some sensitive information, such as a password. When secret starts with <code>file://</code>, the rest of the string is interpreted as a path to a file containing the secret itself: whole content of the file except any trailing whitespace characters is considered a secret value. Note: when clustered, all EMQX nodes should have the same file present before using such config values."""
```

File: apps/emqx_conf/src/emqx_conf.erl [hunk 5]
```diff
@@ -314,18 +316,17 @@ gen_flat_doc(RootNames, #{full_name := FullName, fields := Fields} = S) ->
...
-    [format_field(Field) | format_fields(Fields)].
+format_fields(Fields, DescResolver) ->
+    [format_field(F, DescResolver) || F <- Fields].
 
-format_field(#{name := Name, aliases := Aliases, type := Type} = F) ->
+format_field(#{name := Name, aliases := Aliases, type := Type} = F, DescResolver) ->
     L = [
         {text, Name},
         {type, format_type(Type)},
+        {typedoc, format_type_desc(Type, DescResolver)},
```
Comment (zmstone) [inline apps/emqx_conf/src/emqx_conf.erl:329]: For this version, maybe we can append `dos` after `typedoc`  (insert a `<br/>` in between)

File: apps/emqx_conf/src/emqx_conf_schema_types.erl (hunk 2 above)
Comment (keynslug) [reply apps/emqx_conf/src/emqx_conf_schema_types.erl:194]: There's a small issue with that: users are allowed to write, for instance, `900ms` for the `duration_s()` value, but the result may be very unexpected for them if they don't know the precision. OTOH no similar issue with `duration_ms()`, it's just `Duration` in docs.

File: apps/emqx_conf/src/emqx_conf.erl (hunk 5 above)
Comment (zmstone) [reply apps/emqx_conf/src/emqx_conf.erl:329]: i mean, don't introduce a new field.

Comment (keynslug) [reply apps/emqx_conf/src/emqx_conf.erl:329]: I was under impression that such manipulations should be under the docsite team responsibility.

File: apps/emqx_conf/src/emqx_conf_schema_types.erl (hunk 2 above)
Comment (zmstone) [reply apps/emqx_conf/src/emqx_conf_schema_types.erl:194]: good point

**Review by zmstone — APPROVED** [review]

**Pull request merged**